plasma_cash/root_chain/contracts/RootChain/RootChain.sol

```diff
@@ -102,8 +118,66 @@ contract RootChain {
         require(prevMerkleHash.checkMembership(prevTxObj.uid, prevRoot, prevTxProof));
         require(merkleHash.checkMembership(txObj.uid, root, txProof));
 
-        // Record the exitable timestamp.
-        require(exits[txObj.uid] == 0);
-        exits[txObj.uid] = block.timestamp + 2 weeks;
+        // Record the exit tx.
+        require(!exits[txObj.uid].hasValue);
+        exits[txObj.uid] = exit({
+            hasValue: true,
+            exitTime: block.timestamp + 2 weeks,
+            exitTxBlkNum: txBlkNum,
+            exitTx: tx,
+            txBeforeExitTxBlkNum: prevTxBlkNum,
+            txBeforeExitTx: prevTx
+        });
+    }
+
+    function challengeExit(uint uid, bytes challengeTx, bytes proof, uint blkNum) public {
+        require(exits[uid].hasValue);
+
+        Transaction.Tx memory exitTxObj = (exits[uid].exitTx).createTx();
+        Transaction.Tx memory txBeforeExitTxObj = (exits[uid].txBeforeExitTx).createTx();
+        Transaction.Tx memory challengeTxObj = challengeTx.createTx();
+
+        require(exitTxObj.uid == challengeTxObj.uid);
+        require(exitTxObj.amount == challengeTxObj.amount);
+
+        bytes32 merkleHash = sha3(challengeTx);
+        bytes32 root = childChain[blkNum];
+        require(merkleHash.checkMembership(uid, root, proof));
+
+        if (exitTxObj.newOwner == challengeTxObj.signer) {
+            // Challenge tx spent the exit tx. Cancel it.
+            delete exits[uid].hasValue;
+        } else if (blkNum < exits[uid].exitTxBlkNum
+            && txBeforeExitTxObj.newOwner == challengeTxObj.signer) {
+            // Exit tx double spent the previous tx. Cancel it.
+            delete exits[uid].hasValue;
+        } else if (blkNum < exits[uid].txBeforeExitTxBlkNum) {
+            // Challenger provides a tx in history. Exitor needs to respond it.
+            challenges[uid] = challenge({
```

Comment:
Yes! You are right. I did think of this situation. My solution is that extend the data structure to be a list so there could be many challenges for one exit and all of these should be responded in the end. However, I think it might be too complex for one PR. I intended to finish the modification in the next PR. How do you think? lol

Reply:
Maybe we should leave a issue to this if I want to solve this in the next PR?